core/domain/blog_services.py

```diff
@@ -0,0 +1,613 @@
+
+    if model is None:
+        return None
+
+    return get_blog_post_rights_from_model(model)
+
+
+def get_blog_post_rights_by_user_id(user_id):
+    """Retrieves the rights object for all blog posts for which the given
+    user is an editor.
+
+    Args:
+        user_id: str. ID of the user.
+
+    Returns:
+        list(BlogPostRights). The rights objects associated with the
+        blog posts assigned to given user.
+    """
+    blog_post_rights_models = (
+        blog_models.BlogPostRightsModel.get_by_user(user_id))
+    return [
+        get_blog_post_rights_from_model(model)
+        for model in blog_post_rights_models
+        if model is not None]
+
+
+def does_blog_post_with_url_fragment_exist(url_fragment):
+    """Checks if blog post with provided url fragment exists.
+
+    Args:
+        url_fragment: str. The url fragment for the blog post.
+
+    Returns:
+        bool. Whether the the url fragment for the blog post exists.
+
+    Raises:
+        Exception. Blog Post URL fragment is not a string.
+    """
+    if not isinstance(url_fragment, python_utils.BASESTRING):
+        raise utils.ValidationError(
+            'Blog Post URL fragment should be a string. Recieved:'
+            '%s' % url_fragment)
+    existing_blog_post = get_blog_post_by_url_fragment(url_fragment)
+    return existing_blog_post is not None
+
+
+def _save_blog_post(blog_post):
+    """Saves a BlogPost domain object to the datastore.
+
+    Args:
+        blog_post: BlogPost. The blog post domain object for the given
+            blog post.
+    """
+    model = blog_models.BlogPostModel.get(blog_post.id, strict=True)
+    blog_post.validate()
+
+    model.title = blog_post.title
+    model.content = blog_post.content
+    model.tags = blog_post.tags
+    model.published_on = blog_post.published_on
+    model.thumbnail_filename = blog_post.thumbnail_filename
+    model.url_fragment = blog_post.url_fragment
+    model.update_timestamps()
+    model.put()
+
+
+def publish_blog_post(blog_post_id):
+    """Marks the given blog post as published.
+
+    Args:
+        blog_post_id: str. The ID of the given blog post.
+
+    Raises:
+        Exception. The given blog post does not exist.
+    """
+    blog_post_rights = get_blog_post_rights(blog_post_id, strict=False)
+    if blog_post_rights is None:
+        raise Exception('The given blog post does not exist')
+    blog_post = get_blog_post_by_id(blog_post_id, strict=True)
+    blog_post.validate(strict=True)
+    blog_post_summary = get_blog_post_summary_by_id(blog_post_id, strict=True)
+    blog_post_summary.validate(strict=True)
+    blog_post_rights.blog_post_is_published = True
+
+    published_on = datetime.datetime.utcnow()
+    blog_post.published_on = published_on
+    blog_post_summary.published_on = published_on
+
+    save_blog_post_rights(blog_post_rights)
+    _save_blog_post_summary(blog_post_summary)
+    _save_blog_post(blog_post)
```

Comment:
Should we update the summary and rights inside the `_save_blog_post` function? (This will help us ensure that the summary and rights are always updated!)

Also, can we do the datastore put the call in transaction? [Optional]

Reply:
`save_blog_post_rights` will be called when the blog post is either published or is unpublished. If a blog post is already a draft we will save the blog post and summary and no changes in the rights will be made.

Also we cannot save the blog post summary directly if the blog post is updated, we will need to generate the summary in that case ( but we don't want to do that in case it is simply published without making any changes)

Reply:
Thanks for the info!